src/nvim/os/env.c

```diff
@@ -346,21 +354,47 @@ void init_homedir(void)
   }
 #endif
 
-  if (var != NULL) {
 #ifdef UNIX
+  if (var == NULL) {
+    var = os_homedir();
+  }
+
+  if (var != NULL) {
     // Change to the directory and get the actual path.  This resolves
     // links.  Don't do it when we can't return.
-    if (os_dirname((char_u *)os_buf, MAXPATHL) == OK && os_chdir(os_buf) == 0) {
-      if (!os_chdir(var) && os_dirname(IObuff, IOSIZE) == OK) {
+    if (os_dirname((char_u *)os_buf, sizeof(os_buf)) == OK
+        && os_chdir(os_buf) == 0) {
+      if (!os_chdir(var) && os_dirname(IObuff, sizeof(IObuff)) == OK) {
         var = (char *)IObuff;
       }
       if (os_chdir(os_buf) != 0) {
         EMSG(_(e_prev_dir));
       }
     }
+  }
 #endif
-    homedir = xstrdup(var);
+
+  if ((var == NULL || *var == NUL)
+      && os_dirname((char_u *)os_buf, sizeof(os_buf)) == OK) {
+    var = os_buf;
   }
+
+  homedir = xstrdup(var);
+}
+
+static char homedir_buf[MAXPATHL];
+
+char *os_homedir(void)
+{
+  uv_mutex_lock(&mutex);
+  size_t homedir_size = sizeof(os_buf);
+  if (uv_os_homedir(os_buf, &homedir_size) == 0) {
+    xstrlcpy(homedir_buf, os_buf, homedir_size);
+    uv_mutex_unlock(&mutex);
+    return homedir_buf;
+  }
```

Comment:
Let's `ELOG()` (but not `emsgf()`) the error from `uv_os_homedir`, as in https://github.com/neovim/neovim/pull/9745 .